### PR TITLE
[docs] Add and improve SSL/TLS documentation

### PR DESCRIPTION
**Pull request by bmorelli25**
Closes https://github.com/elastic/apm-server/issues/1579.
**A preview of the changes made in this PR is available [here](http://apm-server_2715.docs-preview.app.elstc.co/guide/en/apm/server/master/securing-apm-server.html)**.

I'm beginning to remember why this issue has been open for almost a year. We are at the mercy of Beats formatting of the `Securing {beatname_uc}` section. I'm not sure these changes would've been possible before Asciidoctor.

This PR adds quite a bit of complexity to the shared Beats files. It's unfortunate, but we've needed this documentation for a long time, so it's probably necessary.

The proposed layout is:
* Configuring APM Server
  * ...
  * SSL/TLS settings
    * SSL output settings
    * SSL input settings
  * ...
* Securing APM Server
  * Secure communication with APM Agents
  * Secure communication with Elasticsearch
  * Secure communication with Logstash by using SSL
  * Use X-Pack security
  * Use Linux Secure Computing Mode (seccomp)

Changes in this PR also need to be persisted in the Beats repo.

### CODE REVIEW

**Comment by bmorelli25** [comment]
Alrighty, here's a rough draft. It'd be great to get a review from @simitt when you have some time. @axw, I've tagged you as well since you opened the issue. If you have time, feel free to leave your thoughts too!

**Comment by axw** [comment]
Thank you @bmorelli25!

A few preliminary thoughts:
 - we have "SSL output settings" at the same level as "Secure communication with Elasticsearch" and "Secure communication with Logstash by using SSL", so it's not really clear where I should get my info
 - likewise "SSL input settings" and "Secure communication with APM Agents"
 - The "by using SSL" in "Secure communication with Logstash by using SSL" is a bit incongruous with Elasticsearch/APM Agents

Are those fixable, or are they the things for which we're at the mercy of Beats?

I'm not 100% on this, but would it be better if "SSL output settings" and "SSL input settings" were under "Configuring APM Server", since they're fairly low-level descriptions of the config attributes? That would leave "Securing APM Server" for higher level guides. We could then cross-reference them.

**Comment by bmorelli25** [comment]
> The "by using SSL" in "Secure communication with Logstash by using SSL" is a bit incongruous with Elasticsearch/APM Agents

I'm not sure why Beats chose this naming convention. You're right that it doesn't fit with the other two. I'll propose the change in Beats!

> * we have "SSL output settings" at the same level as "Secure communication with Elasticsearch" and "Secure communication with Logstash by using SSL", so it's not really clear where I should get my info
> * likewise "SSL input settings" and "Secure communication with APM Agents"
>
> would it be better if "SSL output settings" and "SSL input settings" were under "Configuring APM Server"

This is a good point. My thought process was that having everything under the securing section would make the most sense -- one stop shopping if you will. But you're right, with the heading levels it is a bit confusing to figure out where to start. Let me try moving these two files to the "Configuring APM Server" section and we can see how that looks.

File: docs/secure-communication-agents.asciidoc [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+output.elasticsearch.ssl.certificate_authorities: ["/etc/pki/root/ca.pem"]
+----
+
+A full list of configuration options is available in <<agent-server-ssl>>.
+
+[[https-in-agents]]
+[float]
+=== HTTPS communication in APM Agents
+
+To enable secure communication in your Agents, you simply need to update the configured server URL to use `HTTPS` instead of `HTTP`.
```
Comment (simitt) [inline docs/secure-communication-agents.asciidoc]: I'd remove _simply_, as the more problematic part comes with the CA config etc.

File: docs/secure-communication-agents.asciidoc [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+* *Go Agent*: {apm-go-ref}/configuration.html#config-server-url[`ELASTIC_APM_SERVER_URL`]
+* *Java Agent*: {apm-java-ref}/config-reporter.html#config-server-urls[`server_urls`]
+* *.NET Agent*: {apm-dotnet-ref}/config-reporter.html#config-server-urls[`ServerUrls`]
+* *Node.js Agent*: {apm-node-ref}/configuration.html#server-url[`serverUrl`]
+* *Python Agent*: {apm-py-ref}/[`server_url`]
+* *Ruby Agent*: {apm-ruby-ref}/configuration.html#config-server-url[`server_url`]
+
+Some Agents also allow you to specify a custom certificate authority for connecting to APM Server.
+This allows you to enable encryption and verify that you are sending data to the correct APM Server.
```
Comment (simitt) [inline docs/secure-communication-agents.asciidoc]: > This allows you to enable encryption and verify that you are sending data to the correct APM Server.

I would remove this sentence. You can still verify the server when not configuring a custom CA but using a trusted one, but only for certs signed by a trusted CA. The custom CA config allows you to use and verify self signed certs.

File: docs/ssl-input-settings.asciidoc [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+The `enabled` setting can be used to enable the ssl configuration by setting
+it to `true`. The default value is `false`.
+
+[float]
+==== `certificate_authorities`
+
+The list of root certificates for verifying client certificates.
+If `certificate_authorities` is empty or not set, the trusted certificate authorities of the host system are used.
+Required if `apm-server.enabled` is `true`.
```
Comment (simitt) [inline docs/ssl-input-settings.asciidoc]: It is optional (not required).

File: docs/configuring.asciidoc [hunk 4]
```diff
@@ -25,7 +25,28 @@ include::./copied-from-beats/outputconfig.asciidoc[]
 
 include::./configuring-ingest.asciidoc[]
 
+[[configuration-ssl-landing]]
```
Comment (axw) [inline docs/configuring.asciidoc]: I'm not seeing this in http://apm-server_2715.docs-preview.app.elstc.co/guide/en/apm/server/master/configuring-howto-apm-server.html -- it's still showing "SSL output settings". I suppose the list above needs to be updated.

Comment (bmorelli25) [reply docs/configuring.asciidoc]: Oops, thanks!

File: docs/secure-communication-agents.asciidoc [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+`apm-server.ssl.key` and `apm-server.ssl.certificate` respectively.
+This will make APM Server serve HTTPS requests instead of HTTP.
+
+Example of a basic output config with secure communication enabled:
+
+[source,yaml]
+----
+apm-server.ssl.key: "/etc/pki/key.pem"
+apm-server.ssl.certificate: "/etc/pki/client.pem"
+output.elasticsearch.ssl.certificate_authorities: ["/etc/pki/root/ca.pem"]
```
[the remark this answers is not in the document]
Comment (bmorelli25) [reply docs/secure-communication-agents.asciidoc]: Ahh, copy pasta mistake

File: docs/ssl-input-settings.asciidoc [hunk 6]
```diff
@@ -0,0 +1,79 @@
+[[agent-server-ssl]]
+== SSL input settings
```
[the remark this answers is not in the document]
Comment (bmorelli25) [reply docs/ssl-input-settings.asciidoc]: Can you expand on this? Are you saying you think these settings should _not_ be a separate section, and should instead live on the "[Secure communication with APM Agents](http://localhost:8000/guide/secure-communication-agents.html#ssl-setup)" page?

I like the idea of keeping these settings separate. Right now the "Secure communication with APM Agents" is really all you need to get started with SSL between APM Server/Agents, then users can travel to the input settings page for more advanced information. Like Andrew suggested, it also maintains the current format of all reference config documentation living in the "Configuring APM Server" section.

I could update the sentence at the bottom of [this](http://localhost:8000/guide/secure-communication-agents.html#ssl-setup) section to have more information? Right now it only states "A full list of configuration options is available in SSL input settings."

**Comment by bmorelli25** [comment]
Thanks for the feedback @axw and @simitt!

Silvia, sorry, a lot of those were dumb mistakes. Thanks for catching them 😇. I have one clarifying question above.

> It is hard to understand for APM Server as the mentioned server is the one APM Server talks to. I suggest to change Beat to APM Server:

Good spot -- updated to `{beatname_uc}`.

File: docs/secure-communication-agents.asciidoc [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+* *Node.js Agent*: {apm-node-ref}/configuration.html#secret-token[`Secret Token`]
+* *Python Agent*: {apm-py-ref}/configuration.html#config-secret-token[`secret_token`]
+* *Ruby Agent*: {apm-ruby-ref}/configuration.html#config-secret-token[`secret_token`]
+
+[[ssl-setup]]
+[float]
+=== SSL/TLS communication in APM Server
+
+To enable SSL/TLS, you need to enable SSL and provide both a private key and a certificate
+issued by a certification authority (CA).
```
Comment (axw) [inline docs/secure-communication-agents.asciidoc]: We use "certificate authority" in the other existing pages (as well as below on this page!), should probably change this to be consistent? I think "certificate authority" is more typical anyway.

**Comment by bmorelli25** [comment]
jenkins, run tests again please

**Comment by bmorelli25** [comment]
@elasticmachine, run elasticsearch-ci/docs rebuild

**Comment by bmorelli25** [comment]
jenkins, run tests again please

**Comment by bmorelli25** [comment]
Added to https://github.com/elastic/beats/pull/13596 to persist the shared changes in the Beats repo.

**Pull request merged**